### PR TITLE
adding pie, sunburst, funnel and funnelarea to px

### PR DESCRIPTION
**Pull request by emmanuelle**

### CODE REVIEW

File: packages/python/plotly/plotly/express/_core.py [hunk 1]
```diff
@@ -305,11 +305,38 @@ def make_trace_kwargs(args, trace_spec, g, mapping_labels, sizeref):
             elif k == "locations":
                 result[k] = g[v]
                 mapping_labels[v_label] = "%{location}"
+            elif k == "values":
+                result[k] = g[v]
+                _label = "value" if v_label == "values" else v_label
+                mapping_labels[_label] = "%{value}"
+            elif k == "parents":
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/express/_core.py]: if we want to remove parents and ids from the hover template we can do it here but we could also keep them

File: packages/python/plotly/plotly/express/_doc.py [hunk 2]
```diff
@@ -442,21 +462,41 @@
...
+        "int",
+        "Positive integer",
+        "Sets the number of rendered sectors from any given `level`. Set `maxdepth` to -1 to render all the"
+        "levels in the hierarchy.",
+    ],
 )
 
 
-def make_docstring(fn):
+def make_docstring(fn, override_dict={}):
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/express/_doc.py]: 👍

File: packages/python/plotly/plotly/express/_chart_types.py [hunk 3]
```diff
@@ -1115,3 +1115,172 @@ def parallel_categories(
...
+    labels={},
+    title=None,
+    template=None,
+    width=None,
+    height=None,
+):
+    """
+    In a funnel area plot, each row of `data_frame` is represented as a trapezoidal sector of a funnel.
+    """
+    return make_figure(args=locals(), constructor=go.Funnelarea,)
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/express/_chart_types.py]: here we should force `showlegend=True` just like `pie`

File: packages/python/plotly/plotly/express/_chart_types.py [hunk 4]
```diff
@@ -1115,3 +1115,172 @@ def parallel_categories(
...
+        trace_patch=dict(branchvalues=branchvalues, maxdepth=maxdepth),
+    )
+
+
+treemap.__doc__ = make_docstring(treemap)
+
+
+def funnel(
+    data_frame=None,
+    x=None,
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/express/_chart_types.py]: can we add all the other applicable kwargs from `bar` like `facet_*` and `color_continuous_sequence` etc?

File: packages/python/plotly/plotly/express/_chart_types.py [hunk 5]
```diff
@@ -1115,3 +1115,172 @@ def parallel_categories(
...
+        args=locals(),
+        constructor=go.Funnel,
+        trace_patch=dict(opacity=opacity, orientation=orientation, textinfo=textinfo),
+    )
+
+
+funnel.__doc__ = make_docstring(
+    funnel,
+    override_dict=dict(
+        textinfo=[
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/express/_chart_types.py]: I hesitate about exposing `textinfo` in PX... `texttemplate` is basically the new and improved way to do this, and I think if we're going to expose one or the other, it should be `texttemplate` and it should be probably be added everywhere that text is supported. My feeling is that we should leave out `textinfo` for now and do the broader change in a separate PR.

File: packages/python/plotly/plotly/express/_core.py (hunk 1 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: This makes sense to me ATM.

**Comment by nicolaskruchten** [comment]
This is all pretty good, modulo inline comments.

I'd really like to have support for `color` in pie/funnel/treemap/sunburst the day we release this though... I feel like adding continuous-color support for treemap/sunburst should be pretty trivial (i.e. it should "just work") and then discrete-color support will be harder because for these 4 traces it doesn't work at all the way it does in the other traces: it's not "groupable". But we should try to do it anyway :)

**Comment by nicolaskruchten** [comment]
> I'd really like to have support for color in pie/funnel/treemap/sunburst the day we release this

This would basically be the "value added" of PX for these traces IMO, as compared to using the graph objects directly, so that's why I think it's important :)

**Comment by nicolaskruchten** [comment]
The way I would implement discrete color is basically in  `make_trace_kwargs` with another gross type-specific exception inside the `elif k == "color":` block, which would just hardcode the marker colors.

**Comment by nicolaskruchten** [comment]
Extra requirement: if `color_discrete_sequence` is passed into pie/funnelarea/treemap/sunburst, we should probably assign it to `layout.<whatever>colorway`, so that something like `px.pie(values=[1,2,3], color_discrete_sequence=["red", "green", "blue"])` will do the expected thing.

File: packages/python/plotly/plotly/express/_chart_types.py (hunk 3 above)
Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_chart_types.py]: addressed

File: packages/python/plotly/plotly/express/_chart_types.py (hunk 4 above)
Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_chart_types.py]: done

File: packages/python/plotly/plotly/express/_chart_types.py (hunk 5 above)
Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_chart_types.py]: ok, so I did not put anything and we'll add texttemplate whereever it's relevant in another PR

**Comment by emmanuelle** [comment]
Hum, interesting to see that the test fail because the color of the funnel is not the same.

**Comment by emmanuelle** [comment]
Do we want to accept `color_discrete_sequence='Viridis'`? At the moment it only works with a `px` object.

**Comment by nicolaskruchten** [comment]
it's normal that `color_discrete_sequence` only works with an array of colors, yes. It's not clear what we would do if we got `Viridis`, as that scale doesn't have a fixed number of categorical colors...?

**Comment by emmanuelle** [comment]
ok good, so nothing more to do for this :-).

**Comment by nicolaskruchten** [comment]
@emmanuelle some tweaks in that last commit, LMK what you think.

File: packages/python/plotly/plotly/express/_core.py [hunk 6]
```diff
@@ -304,26 +291,28 @@ def make_trace_kwargs(args, trace_spec, g, mapping_labels, sizeref):
...
-                        g[v], args["color_discrete_sequence"]
-                    )
+                        result["marker"]["colors"] = []
+                        mapping = {}
+                        for cat in g[v]:
+                            if mapping.get(cat) is None:
+                                mapping[cat] = args["color_discrete_sequence"][
+                                    len(mapping) % len(args["color_discrete_sequence"])
+                                ]
+                            result["marker"]["colors"].append(mapping[cat])
```
Comment (emmanuelle) [inline packages/python/plotly/plotly/express/_core.py]: Oh interesting, I did not know that `marker_colors` could be a dictionary, I thought it had to be a list/array.

**Comment by emmanuelle** [comment]
Thanks, look good to me. Should we give each other a :dancer: ? :-)

File: packages/python/plotly/plotly/express/_core.py (hunk 6 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: I think it does have to be a list, and in this case it is...?

Comment (nicolaskruchten) [reply packages/python/plotly/plotly/express/_core.py]: this is the same code you had, just inlined :)

Comment (emmanuelle) [reply packages/python/plotly/plotly/express/_core.py]: oh yes I had read too fast!
